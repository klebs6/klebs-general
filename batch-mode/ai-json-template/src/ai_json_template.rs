// ---------------- [ File: ai-json-template/src/ai_json_template.rs ]
crate::ix!();

/// The derived code implements `AiJsonTemplate` for each struct, letting you
/// call `MyStruct::to_template()` to get a JSON “schema” describing how the
/// AI should produce data that matches this layout.
///
pub trait AiJsonTemplate
: Clone 
+ Debug 
+ LoadFromFile 
+ SaveToFile<Error=SaveLoadError> 
+ Serialize 
+ for<'a> Deserialize<'a> 
{
    /// Return a JSON template describing how the AI’s output should be structured.
    /// This might include doc comments or other instructions for each field.
    fn to_template() -> serde_json::Value;
}
