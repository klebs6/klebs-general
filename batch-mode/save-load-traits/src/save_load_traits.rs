// ---------------- [ File: src/save_load_traits.rs ]
crate::ix!();

#[async_trait]
pub trait SaveToFile {

    type Error;

    async fn save_to_file(
        &self,
        filename: impl AsRef<Path> + Send,

    ) -> Result<(), Self::Error>;
}

#[async_trait]
pub trait LoadFromFile: Sized {

    type Error;

    async fn load_from_file(filename: impl AsRef<Path> + Send) 
        -> Result<Self, Self::Error>;

}
