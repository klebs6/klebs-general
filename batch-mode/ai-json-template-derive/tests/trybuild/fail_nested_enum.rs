// ---------------- [ File: ai-json-template-derive/tests/trybuild/fail_nested_enum.rs ]
// ======================= File: tests/trybuild/fail_nested_enum.rs =======================
// Tests a struct that nests an enum. The macro's logic (in this codebase) doesn't handle enums,
// so it should fail.

#![allow(dead_code)]

use ai_json_template_derive::*;
use ai_json_template::*;
use serde::{Serialize, Deserialize};
use getset::{Getters, Setters};
use derive_builder::Builder;
use save_load_derive::*;
use save_load_traits::*;

#[derive(SaveLoad,Clone, Debug, Serialize, Deserialize)]
enum InnerEnum {
    VariantA,
    VariantB,
}

#[derive(SaveLoad,Clone, Debug, Serialize, Deserialize, Getters, Setters, Builder)]
#[derive(AiJsonTemplate)]
#[getset(get = "pub", set = "pub")]
#[builder(setter(into))]
struct FailNestedEnum {
    /// The presence of an enum here should cause the derive to fail
    something: InnerEnum,

    /// Some other field
    text: String,
}

fn main() {}
