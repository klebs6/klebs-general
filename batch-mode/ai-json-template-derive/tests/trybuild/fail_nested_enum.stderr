error[E0277]: the trait bound `InnerEnum: ai_json_template::AiJsonTemplate` is not satisfied
  --> tests/trybuild/fail_nested_enum.rs:29:16
   |
29 |     something: InnerEnum,
   |                ^^^^^^^^^ the trait `ai_json_template::AiJsonTemplate` is not implemented for `InnerEnum`
   |
   = help: the following other types implement trait `ai_json_template::AiJsonTemplate`:
             FailNestedEnum
             Vec<std::string::String>
